TACHO_DRIVER.md

```diff
@@ -0,0 +1,175 @@
+
+The `tacho-motor` drivers must strike a balance between doing the most common
+functions that a robot builder needs but not becoming so complex that the
+operation is unreliable or mysterious.
+
+## The sys Interface
+
+All motor operations are contolled by writing data to attributes within the
+`/sys/class/tacho-motor/motorN/` folder. There is one folder per motor.
+
+The read and write operations may return the following return codes:
+
+| RETURN CODE    | DESCRIPTION
+| ------------   | -----------
+| 0              | Operation sucessful
+| -13 EACCES     | Permission denied
+| -16 EBUSY      | Device or resource busy
+| -22 EINVAL     | Invalid argument
+| -95 EOPNOTSUPP | Operation not supported on transport endpoint
+
+Note that some hardware supported by the [ev3dev.org] distribution uses
+I2C or UART communication channels for motor control, and may return 
+additional error codes.
+
+The attributes operate as described in the
+[`tacho-motor` class document](http://www.ev3dev.org/docs/drivers/tacho-motor-class/).
+In most cases, changing attributes after writing to the `command` attribute
+will not take effect until the next write to `command`.
+
+## Changes to the current operation
+
+### `run-direct-duty-cycle` and `run-direct-speed`
+
+The `command` attribute will be updated to replace `run-direct` with
+two new commands that are intended for applications that need more
+advanced control over the motors. This includes "steering" mode control
+where you want motors to operate at a ratio to each other.
+
+The new commands ignore the value of any attribute except for the one
+that they are named after. 
+
+| `command` value         | Controlling attribute
+| ------------------      | ---------------------
+| `run-direct-duty-cycle` | `duty_cycle_sp`
+| `run-direct-speed`      | `speed_sp`
+
+If a motor does not support a particular commands, then the write
+will return `EOPNOTSUPP`.
+
+When these commands are used, there is no support for ramping - it is
+up to the application program to handle any requirements for motion
+control algorithms.
+
+Note that if the motor supports `run-direct-speed` the attributes for
+the PID control do affect speed regulation, but they do not take
+effect until the next write to `command`.
+
+The value of the controlling attribute can be changed at any time and will
+take effect immediately.
+
+When the value of the controlling attribute is set to 0, then the motor
+is stopped according to the value in the `stop_command` attribute.
+
+When `stop` is written to the `command` attribute, then the motor
+is stopped according to the value in the `stop_command` attribute.
+
+A write to any other attribute will not take effect until the next
+write to `command`.
+
+### Ramping Updates
+
+The automatic linear ramping only works when one of the following
+values is written to the `command` attribute:
+
+- `run-forever`
+- `run-timed`
+- `run-to-abs-pos`
+- `run-to-rel-pos`
+- `stop`
+
```

Comment:
I've updated the proposal to clarify the goal - the EV3 and built-in tacho motor hardware are the gold standard. We implement what we already have, including ramping when running to position.

The EV3 hardware allows a simple memory read to get the current tacho position, calculate speed, and read the duty cycle - so it's natural that the tacho motor class can simply issue a call into the motor driver to immediately retreive these values for calculations.

Other hardware like the multiplexers, or shields for rPi or BB may not support such ready access to motor/tacho state. Too bad - they have their own ramping, PID, and so on. If the tacho motor class needs to read the current position and the motor driver can't suppy it, then it returns EBUSY if we need to try again or EOPNOTSUPP if it's impossible.

For those other hardware, you don't get flexible ramping that requires position feedback. But you still get run to position and regular `run-timed` ramping.